File: Othello.cpp
#include <iostream>
#include <vector>
#include <string.h>
#include <windows.h>

using namespace std;

class Piece {
    private:
        int x, y;
        string color;
    public:
        Piece (int x_pos, int y_pos, string c){
            x = x_pos;
            y = y_pos;
            color = c;
        }
        int getx() {return x;}
        int gety() {return y;}
        string getColor() {return color;}
        void setColor(string c) {color = c;}
        void set (int x_pos, int y_pos, string c) {
            x = x_pos;
            y = y_pos;
            color = c;
        }
        void flip() {
            if (color == "black") {color = "white";}
            else {color = "black";}
        }
};

class Board {
    private:
        vector<Piece> list;         ///Vector of existing pieces
        Piece getPiece(int x, int y) {      ///Returns piece at location
            for (int i=0; i<list.size(); i++) {
                if (list.at(i).getx() == x && list.at(i).gety() == y) {return list.at(i);}
            }
        }
        int getIndex(int x, int y) {        ///Returns vector index of piece at location
            for(int i=0; i<list.size(); i++) {
                if (list.at(i).getx() == x && list.at(i).gety() == y) {return i;}
            }
        }
        int getBlackNum () {            ///Returns # of black pieces
            int num = 0;
            for (int i=0; i<list.size(); i++) {
                if (list.at(i).getColor().compare("black") == 0) {num++;}
            }
            return num;
        }
        int getWhiteNum () {            ///Returns # of white pieces
            int num = 0;
            for (int i=0; i<list.size(); i++) {
                if (list.at(i).getColor().compare("white") == 0) {num++;}
            }
            return num;
        }
        void display() {               ///Displays board state
            cout << "   Board" << endl << "   +---+---+---+---+---+---+---+---+" << endl;
            for (int y=8; y>=1; y--) {
                cout << " " << y << " | ";
                for (int x=1; x<=8; x++) {
                    if (pAtLoc(x, y)) {
                        if (getPiece(x, y).getColor()=="black") {cout << "b";}
                        else if (getPiece(x, y).getColor()=="white") {cout << "w";}
                    }
                    else {
                        cout << " ";
                    }
                    cout << " | ";
                }
                cout << endl << "   +---+---+---+---+---+---+---+---+" << endl;
            }
            cout << "     1   2   3   4   5   6   7   8  " << endl << endl;
        }
        bool pAtLoc(int x, int y) {      ///Checks for existence of piece at location
            for (int i=0; i<list.size(); i++) {
                if (list.at(i).getx()==x && list.at(i).gety()==y) {
                    return true;
                }
            }
            return false;
        }
        bool inRange(int x, int y) {     ///Checks if location is on the board
            return !(x<1 || x>8 || y<1 || y>8);
        }
        bool legalDown(int x, int y, string color) {
            int currenty;

            if (pAtLoc(x, y-1)) {
                if (getPiece(x, y-1).getColor().compare(color) != 0) {
                    currenty = y-1;
                    while (pAtLoc(x, currenty-1)) {
                        if (getPiece(x, currenty-1).getColor().compare(color) == 0) {return true;}
                        else {
                            currenty--;
                        }
                    }
                }
            }
            return false;
        }
        bool legalUp(int x, int y, string color) {
            int currenty;

            if (pAtLoc(x, y+1)) {
                if (getPiece(x, y+1).getColor().compare(color) != 0) {
                    currenty = y+1;
                    while (pAtLoc(x, currenty+1)) {
                        if (getPiece(x, currenty+1).getColor().compare(color) == 0) {return true;}
                        else {
                            currenty++;
                        }
                    }
                }
            }
            return false;
        }
        bool legalLeft(int x, int y, string color) {
            int currentx;

            if (pAtLoc(x-1, y)) {
                if (getPiece(x-1, y).getColor().compare(color) != 0) {
                    currentx = x-1;
                    while (pAtLoc(currentx-1, y)) {
                        if (getPiece(currentx-1, y).getColor().compare(color) == 0) {return true;}
                        else {
                            currentx--;
                        }
                    }
                }
            }
            return false;
        }
        bool legalRight(int x, int y, string color) {
            int currentx;

            if (pAtLoc(x+1, y)) {
                if (getPiece(x+1, y).getColor().compare(color) != 0) {
                    currentx = x+1;
                    while (pAtLoc(currentx+1, y)) {
                        if (getPiece(currentx+1, y).getColor().compare(color) == 0) {return true;}
                        else {
                            currentx++;
                        }
                    }
                }
            }
            return false;
        }
        bool legalUpLeft(int x, int y, string color) {
            int currentx, currenty;

            if (pAtLoc(x-1, y+1)) {
                if (getPiece(x-1, y+1).getColor().compare(color) != 0) {
                    currentx = x-1;
                    currenty = y+1;
                    while (pAtLoc(currentx-1, currenty+1)) {
                        if (getPiece(currentx-1, currenty+1).getColor().compare(color) == 0) {return true;}
                        else {
                            currentx--;
                            currenty++;
                        }
                    }
                }
            }
            return false;
        }
        bool legalUpRight(int x, int y, string color) {
            int currentx, currenty;

            if (pAtLoc(x+1, y+1)) {
                if (getPiece(x+1, y+1).getColor().compare(color) != 0) {
                    currentx = x+1;
                    currenty = y+1;
                    while (pAtLoc(currentx+1, currenty+1)) {
                        if (getPiece(currentx+1, currenty+1).getColor().compare(color) == 0) {return true;}
                        else {
                            currentx++;
                            currenty++;
                        }
                    }
                }
            }
            return false;
        }
        bool legalDownRight(int x, int y, string color) {
            int currentx, currenty;

            if (pAtLoc(x+1, y-1)) {
                if (getPiece(x+1, y-1).getColor().compare(color) != 0) {
                    currentx = x+1;
                    currenty = y-1;
                    while (pAtLoc(currentx+1, currenty-1)) {
                        if (getPiece(currentx+1, currenty-1).getColor().compare(color) == 0) {return true;}
                        else {
                            currentx++;
                            currenty--;
                        }
                    }
                }
            }
            return false;
        }
        bool legalDownLeft(int x, int y, string color) {
            int currentx, currenty;

            if (pAtLoc(x-1, y-1)) {
                if (getPiece(x-1, y-1).getColor().compare(color) != 0) {
                    currentx = x-1;
                    currenty = y-1;
                    while (pAtLoc(currentx-1, currenty-1)) {
                        if (getPiece(currentx-1, currenty-1).getColor().compare(color) == 0) {return true;}
                        else {
                            currentx--;
                            currenty--;
                        }
                    }
                }
            }
            return false;
        }
        bool legalMove(int x, int y, string color) {        ///Checks if legal move exists at location for specific color
            if (pAtLoc(x, y)|| !inRange(x, y)) {return false;}
            return legalDown(x, y, color) || legalUp(x, y, color) || legalLeft(x, y, color) || legalRight(x, y, color) || legalUpLeft(x, y, color) || legalUpRight(x, y, color) || legalDownRight(x, y, color) || legalDownLeft(x, y, color);
        }
        void placePiece(int x, int y, string color) {       ///Places piece and flips as needed
            int currentx, currenty;

            if (legalDown(x, y, color)) {
                currenty = y-1;
                while(getPiece(x, currenty).getColor().compare(color) != 0) {
                    list.at(getIndex(x, currenty)).flip();
                    currenty--;
                }
            }
            if(legalUp(x, y, color)) {
                currenty = y+1;
                while(getPiece(x, currenty).getColor().compare(color) != 0) {
                    list.at(getIndex(x, currenty)).flip();
                    currenty++;
                }
            }
            if(legalLeft(x, y, color)) {
                currentx = x-1;
                while(getPiece(currentx, y).getColor().compare(color) != 0) {
                    list.at(getIndex(currentx, y)).flip();
                    currentx--;
                }
            }
            if(legalRight(x, y, color)) {
                currentx = x+1;
                while(getPiece(currentx, y).getColor().compare(color) != 0) {
                    list.at(getIndex(currentx, y)).flip();
                    currentx++;
                }
            }
            if(legalUpLeft(x, y, color)) {
                currentx = x-1;
                currenty = y+1;
                while(getPiece(currentx, currenty).getColor().compare(color) != 0) {
                    list.at(getIndex(currentx, currenty)).flip();
                    currentx--;
                    currenty++;
                }
            }
            if(legalUpRight(x, y, color)) {
                currentx = x+1;
                currenty = y+1;
                while(getPiece(currentx, currenty).getColor().compare(color) != 0) {
                    list.at(getIndex(currentx, currenty)).flip();
                    currentx++;
                    currenty++;
                }
            }
            if(legalDownRight(x, y, color)) {
                currentx = x+1;
                currenty = y-1;
                while(getPiece(currentx, currenty).getColor().compare(color) != 0) {
                    list.at(getIndex(currentx, currenty)).flip();
                    currentx++;
                    currenty--;
                }
            }
            if(legalDownLeft(x, y, color)) {
                currentx = x-1;
                currenty = y-1;
                while(getPiece(currentx, currenty).getColor().compare(color) != 0) {
                    list.at(getIndex(currentx, currenty)).flip();
                    currentx--;
                    currenty--;
                }
            }
            list.push_back(Piece(x, y, color));
        }
        bool checkLegal(string color) {         ///Checks if legal move exists for specific color
            for (int x=1; x<=8; x++) {
                for (int y=1; y<=8; y++) {
                    if (legalMove(x, y, color)) {return true;}
                }
            }
            return false;
        }
    public:
        Board() {
            ///Initial board state
            list.push_back(Piece(4, 4, "black"));
            list.push_back(Piece(5, 5, "black"));
            list.push_back(Piece(4, 5, "white"));
            list.push_back(Piece(5, 4, "white"));
        }
        void execute() {
            int turn = 1;
            int gameOver = 0;
            string col[2] = {"white", "black"};
            do {
                bool first = true;
                int x, y;

                ///Game loop
                while(!legalMove(x, y, col[turn]) || first) {
                    system("cls");
                    display();
                    cout << col[turn] << "'s turn" << endl << endl;
                    if(!legalMove(x, y, col[turn]) && !first) {
                        cout << "Illegal move" << endl;
                    }
                    cout << "X: ";
                    cin >> x;
                    cout << "Y: " ;
                    cin >> y;
                    first = false;
                }
                placePiece(x, y, col[turn]);

                if (checkLegal(col[(turn+1)%2])) {
                    turn = (turn+1)%2;
                }
            } while (checkLegal("white") || checkLegal("black"));

            ///Win Screen
            system("cls");
            display();
            cout << "Game over" << endl;
            cout << "Black pieces: " << getBlackNum() << endl;
            cout << "White pieces: " << getWhiteNum() << endl;
            if (getBlackNum() > getWhiteNum()) {
                cout << "Black wins" << endl;
            } else if (getBlackNum() == getWhiteNum()) {
                cout << "Tie" << endl;
            } else {
                cout << "White wins" << endl;
            }
        }
};

int main() {
    Board b;
    b.execute();

    return 0;
}

File: othello.py
import os
import random
import sys

class Piece:
    def __init__(self, x, y, c):
        self.x, self.y, self.color = x, y, c

    def flip(self):
        if self.color == 'b':
            self.color = 'w'
        else:
            self.color = 'b'

class Board:
    def __init__(self):
        self.pieces = [Piece(4, 4, 'b'), Piece(5, 5, 'b'), Piece(4, 5, 'w'), Piece(5, 4, 'w')]

    def white_num(self):
        count = 0
        for i in self.pieces:
            if i.color == 'w':
                count += 1
        return count

    def black_num(self):
        count = 0
        for i in self.pieces:
            if i.color == 'b':
                count += 1
        return count

    def display(self):
        print("   Board")
        print("   +---+---+---+---+---+---+---+---+")
        for y in range(8, 0, -1):
            print(" " + str(y) + " | ", end = '')
            for x in range(1, 9, 1):
                if self.get_piece(x, y):
                    if self.get_piece(x, y).color == 'b':
                        print('b', end = '')
                    elif self.get_piece(x, y).color == 'w':
                        print('w', end = '')
                else:
                    print (' ', end = '')
                print(" | ", end = '')
            print("\n   +---+---+---+---+---+---+---+---+")
        print("     1   2   3   4   5   6   7   8  ")

    def get_piece(self, x, y):
        for i in self.pieces:
            if i.x == x and i.y == y:
                return i
        return False

    def get_color(self, x, y):
        for i in self.pieces:
            if i.x == x and i.y == y:
                return i.color
        return False

    def in_range(self, x, y):
        return x >= 1 and x <= 8 and y >= 1 and y <= 8

    def check_move(self, x, y, color): # returns list of legal flips
        if self.get_piece(x, y) or not self.in_range(x, y):
            return []

        tiles_to_flip = []
        if color == 'w':
            opp_color = 'b'
        else:
            opp_color = 'w'

        for xdir, ydir in [[0, -1], [0, 1], [-1, 0], [1, 0], [-1, 1], [1, 1], [1, -1], [-1, -1]]:
            xcurr, ycurr, count = x, y, 0
            xcurr += xdir
            ycurr += ydir
            while self.in_range(xcurr, ycurr) and self.get_color(xcurr, ycurr) == opp_color:
                xcurr += xdir
                ycurr += ydir
                count += 1
            if (self.get_color(xcurr, ycurr) == color and count > 0):
                xcurr -= xdir
                ycurr -= ydir
                while xcurr != x or ycurr != y:
                    tiles_to_flip.append([xcurr, ycurr])
                    xcurr -= xdir
                    ycurr -= ydir

        return tiles_to_flip

    def make_move(self, flips, x, y, color): # flips coordinates in flip list
        for i in flips:
            self.get_piece(i[0], i[1]).flip()
        self.pieces.append(Piece(x, y, color))

    def check_legal(self, color): # check_move for all coords of a color
        for x in range(1, 9, 1):
            for y in range(1, 9, 1):
                if self.check_move(x, y, color) != []:
                    return True
        return False

    def execute(self):
        os.system('cls')
        choice = int(input("Choose mode:\n(1)Free Play\n(2)Beginner AI\n(3)Intermediate AI\n"))

        x, y, numFlips, turn = 0, 0, 0, 1
        col = ["w","b"]

        while b.check_legal("w") or b.check_legal("b"):
            flips = []

            if turn % 2 == 1 or choice == 1:
                first = True
                while flips == []:
                    os.system('cls')
                    self.display()

                    if choice == 1 and col[turn % 2] == 'w':
                        print("White's turn\n")
                    elif choice == 1 and col[turn % 2] == 'b':
                        print("Black's turn\n")
                    else:
                        print("Your turn (black)\n")

                    if first and turn != 1 and choice != 1:
                        print("Computer placed a piece at (" + str(x) + ", " + str(y) + ") and flipped " + str(numFlips) + " pieces")
                    if flips == [] and first == False:
                        print("Illegal move\n")

                    x = int(input("X: "))
                    y = int(input("Y: "))
                    flips = b.check_move(x, y, col[turn % 2])
                    print(b.check_legal(col[turn % 2]))
                    first = False
                b.make_move(flips, x, y, col[turn % 2])
            elif choice == 2:
                x, y, numFlips = b.beginner_AI(turn)
            elif choice == 3:
                x, y, numFlips = b.intermediate_AI(turn)

            if b.check_legal(col[(turn + 1) % 2]):
                turn += 1

        os.system('cls')
        b.display()
        print("Game over")
        print("Black pieces: " + str(b.black_num()))
        print("White pieces: " + str(b.white_num()))

        if b.black_num() > b.white_num():
            print("Black wins")
        elif b.white_num() > b.black_num():
            print("White wins")
        else:
            print("Tie")

    def beginner_AI(self, turn):
        col = ['w', 'b']
        while True:
            x = random.randint(1, 8)
            y = random.randint(1, 8)
            flips = b.check_move(x, y, col[turn % 2])
            if flips != []:
                break
        b.make_move(flips, x, y, col[turn % 2])
        return [x, y, len(flips)]

    def intermediate_AI(self, turn): # + randint(0, 6) % 8 + 1
        col = ['w', 'b']
        flips = []
        range = [1, 2, 3, 4, 5, 6, 7, 8]
        random.shuffle(range)
        for x in range:
            for y in range:
                if len(b.check_move(x, y, col[turn % 2])) > len(flips):
                    flips = b.check_move(x, y, col[turn % 2])
                    bestx = x
                    besty = y
        b.make_move(flips, bestx, besty, col[turn % 2])
        return [bestx, besty, len(flips)]


b = Board()
b.execute()
